builder/amazon/common/access_config.go

```diff
@@ -237,6 +305,177 @@ func (c *AccessConfig) IsChinaCloud() bool {
 	return strings.HasPrefix(c.SessionRegion(), "cn-")
 }
 
+// GetCredentials gets credentials from the environment, shared credentials,
+// the session (which may include a credential process), or ECS/EC2 metadata
+// endpoints. GetCredentials also validates the credentials and the ability to
+// assume a role or will return an error if unsuccessful.
+func (c *AccessConfig) GetCredentials(config *aws.Config) (*awsCredentials.Credentials, error) {
```

Comment:
This flow is copied pretty closely from Terraform. I've tested it with all the common use cases (env, credentials file, assume role, hard-coded credentials) and it works as expected.